kernel: Remove protocol.h/netaddress.h/compat.h from kernel headers

sedited opened the pull request:
This removes the non-consensus critical `protocol.h` and `netaddress.h` headers from the kernel headers. With this patch, they are no longer required to include in order to use the libbitcoinkernel library. This also allows for the removal of the `compat.h` header from the kernel headers.

As an added future benefit it also reduces the number of of kernel headers that include the platform specific `bitcoin-config.h`.

For those interested, the currently required kernel headers can be inspected visually with the [sourcetrail](https://github.com/CoatiSoftware/Sourcetrail) tool by looking at the required includes of `bitcoin-chainstate.cpp`.

---

This is part of the [libbitcoinkernel project](https://github.com/bitcoin/bitcoin/issues/27587), namely its stage 1 step 3: Decouple most non-consensus headers from libbitcoinkernel.

Commit:
[refactor] Allow std::array<std::byte, N> in serialize.h

This is already possible for C-style arrays, so allow it for C++11
std::array as well.

sedited committed:
kernel: Move MessageStartChars to its own file

The protocol.h file contains many non-consensus related definitions and
should thus not be part of the libbitcoinkernel. This commit makes
protocol.h no longer a required include for users of the
libbitcoinkernel.

This commit is part of the libbitcoinkernel project, namely its stage 1
step 3: Decouple most non-consensus headers from libbitcoinkernel.

Co-Authored-By: Cory Fields <cory-nospam-@coryfields.com>

sedited committed:
[refactor] Add missing includes for next commit

ajtowns commented on src/messageheader.h:
```diff
@@ -0,0 +1,46 @@
+    static constexpr size_t MESSAGE_SIZE_OFFSET = MESSAGE_START_SIZE + COMMAND_SIZE;
+    static constexpr size_t CHECKSUM_OFFSET = MESSAGE_SIZE_OFFSET + MESSAGE_SIZE_SIZE;
+    static constexpr size_t HEADER_SIZE = MESSAGE_START_SIZE + COMMAND_SIZE + MESSAGE_SIZE_SIZE + CHECKSUM_SIZE;
+    typedef unsigned char MessageStartChars[MESSAGE_START_SIZE];
```
Would it make more sense to pull `MessageStartChars` alone into kernel/chainparams.h, and just add a `static_assert(sizeof(MessageStartChars) == MESSAGE_START_SIZE);` here?

TheCharlatan replied:
Yes, that works too and I went through such an iteration. I thought adding such logic might be a bit less desirable though than just straight up moving these simple definitions and keeping the lines that are actually changed to ~0, especially since the kernel code uses definitions to both `MessageStartChars` and its size.

ajtowns replied:
I guess I look at it more as "chainparams provides a 4 byte magic number" and "other things (p2p, wallet, block store) use those magic numbers to help stop you using a different chain than you intended to". That seems like a sensible thing for chainparams to do; but it's just that one line, all the other stuff in this header is p2p protocol things which don't seem interesting for the kernel?

TheCharlatan replied:
I don't mind either way, this is a simple change. To be clear, is your intention: Move `MESSAGE_START_SIZE` and the `MessageStartChars` to `kernel/chainparams.h`, then include `kernel/chainparams.h` as needed?

theuni replied:
I believe the two approaches work just fine together? https://github.com/theuni/bitcoin/commit/23b992e661650e87f572895ee4bad0e6e5e5196a

ajtowns replied:
What @theuni said.

TheCharlatan replied:
> If you take that, please use less stupid names than I did :p

Pretty sure I won't :P

BlazerX98 commented:
Ok.Approved in it...

Regards
Prince Johnso

On Thu, Sep 7, 2023, 10:55 AM Reproducibility Matters <
***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/messageheader.h
> <https://github.com/bitcoin/bitcoin/pull/28423#discussion_r1318096043>:
>
> > + * (4) message start.
> + * (12) command.
> + * (4) size.
> + * (4) checksum.
> + */
> +class CMessageHeader
> +{
> +public:
> +    static constexpr size_t MESSAGE_START_SIZE = 4;
> +    static constexpr size_t COMMAND_SIZE = 12;
> +    static constexpr size_t MESSAGE_SIZE_SIZE = 4;
> +    static constexpr size_t CHECKSUM_SIZE = 4;
> +    static constexpr size_t MESSAGE_SIZE_OFFSET = MESSAGE_START_SIZE + COMMAND_SIZE;
> +    static constexpr size_t CHECKSUM_OFFSET = MESSAGE_SIZE_OFFSET + MESSAGE_SIZE_SIZE;
> +    static constexpr size_t HEADER_SIZE = MESSAGE_START_SIZE + COMMAND_SIZE + MESSAGE_SIZE_SIZE + CHECKSUM_SIZE;
> +    typedef unsigned char MessageStartChars[MESSAGE_START_SIZE];
>
> Yes, that works too and I went through such an iteration. I thought adding
> such logic might be a bit less desirable though than just straight up
> moving these simple definitions and keeping the lines that are actually
> changed to ~0, especially since the kernel code uses definitions to both
> MessageStartChars and its size.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/28423#discussion_r1318096043>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJTHIURGYRRLS3E6SWXSZMLXZFLDXANCNFSM6AAAAAA4N3DJPA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

sedited committed:
[refactor] Remove compat.h from kernel headers

This commit makes compat.h no longer a required include for users of the
libbitcoinkernel. Including compat.h imports a bunch of
platform-specific definitions.

This commit is part of the libbitcoinkernel project, namely its stage 1
step 3: Decouple most non-consensus headers from libbitcoinkernel.

theuni commented:
> This removes the non-consensus critical protocol.h and netaddress.h headers from the kernel headers. 

Very nice. Concept ACK.

TheCharlatan commented:
Updated d4cc87f9624f38e90698f320f427204d504e4d36 -> 5bebabda28f10e6069e1d0dbb2828ed9db2a5696 ([rmNetKernel_0](https://github.com/TheCharlatan/bitcoin/tree/rmNetKernel_0) -> [rmNetKernel_1](https://github.com/TheCharlatan/bitcoin/tree/rmNetKernel_1),  [compare](https://github.com/TheCharlatan/bitcoin/compare/rmNetKernel_0..rmNetKernel_1))

* Addressed @ajtowns's [comment](https://github.com/bitcoin/bitcoin/pull/28423#discussion_r1317918689), moving only the required typedef and constant to the kernel includes now.
* Reverted the creation of `messageheader.h`. With the current patch I am not sure what utility this would still provide.

theuni reviewed (approved):
ACK 978baafe99bdd699940e305cd5501b4e21409ca3

I didn't mind the additional messageheader.h split before, but without it this way is fine too.

The `GetDefaultPort()`'s are awkward in their new home, but I don't have a better suggestion.

ajtowns reviewed (commented):
ACK 978baafe99bdd699940e305cd5501b4e21409ca3

TheCharlatan commented:
Updated 978baafe99bdd699940e305cd5501b4e21409ca3 -> 197de6a98657bf1279ac22aa38cde8dc79129088 ([rmNetKernel_1](https://github.com/TheCharlatan/bitcoin/tree/rmNetKernel_1) -> [rmNetKernel_2](https://github.com/TheCharlatan/bitcoin/tree/rmNetKernel_2),  [compare](https://github.com/TheCharlatan/bitcoin/compare/rmNetKernel_1..rmNetKernel_2))

* Addressed @ajtowns's [comment](https://github.com/bitcoin/bitcoin/pull/28423#discussion_r1317918689), moving `GetDefaultPort` functions to `CConnman` instead.

TheCharlatan commented on src/netaddress.cpp:
```diff
@@ -20,6 +21,17 @@
 #include <iterator>
 #include <tuple>
 
+uint16_t GetDefaultPort(const CChainParams& params, Network net)
```
Thanks, that makes more sense, yes.

ajtowns commented:
ACK 197de6a98657bf1279ac22aa38cde8dc79129088

theuni reviewed (approved):
ACK 197de6a98657bf1279ac22aa38cde8dc79129088.

I would've preferred to see the Params change in `CConnman` split into a separate commit, as the implications of that aren't 100% clear to me (Mostly: can we trickle `m_params` down anywhere else now?).

But it looks correct and is an improvement on top of the original PR.

maflcko reviewed (commented):
lgtm, left a nit

theuni commented on src/util/time.h:
```diff
@@ -6,8 +6,6 @@
 #ifndef BITCOIN_UTIL_TIME_H
 #define BITCOIN_UTIL_TIME_H
 
-#include <compat/compat.h>
```
Ah yes, nice catch. `MillisToTimeval()` is the problem here. Ideally that should be moved away as `timeval` isn't api safe.

stickies-v reviewed (approved):
ACK 197de6a98657bf1279ac22aa38cde8dc79129088

Not super happy about the `GetDefaultPort` situation, but I can't think of a better solution in this PR either.

TheCharlatan commented:
Thank you for the reviews, will try to address the remaining nits.

TheCharlatan commented on src/kernel/messagestartmagic.h:
```diff
@@ -0,0 +1,15 @@
+
+#include <cstddef>
+
+namespace MessageStartMagic {
+static constexpr size_t MESSAGE_START_SIZE = 4;
+typedef unsigned char MessageStartChars[MESSAGE_START_SIZE];
+} // namespace MessageStartMagic
```
This looks like a larger refactor to me now that I have implemented this suggestion. Is there an easy way to make `std::array<uint8_t, 4>` behave with our serialization?

hebasto commented:
Concept ACK.

hebasto reviewed (approved):
Approach ACK 197de6a98657bf1279ac22aa38cde8dc79129088.

sedited committed:
[refactor] Define MessageStartChars as std::array

sedited committed:
[refactor] Add CChainParams member to CConnman

This is done in preparation to the next commit, but has the nice
effect of removing one further data structure relying on the global
`Params()`.

sedited committed:
[refactor] Remove netaddress.h from kernel headers

Move functions requiring the netaddress.h include out of
libbitcoinkernel source files.

The netaddress.h file contains many non-consensus related definitions
and should thus not be part of the libbitcoinkernel. This commit makes
netaddress.h no longer a required include for users of the
libbitcoinkernel.

This commit is part of the libbitcoinkernel project, namely its stage 1
step 3: Decouple most non-consensus headers from libbitcoinkernel.

TheCharlatan commented:
Updated 197de6a98657bf1279ac22aa38cde8dc79129088 -> 6ec2d752e9edf83dd9b2b7f792dbbd17026b3aa5 ([rmNetKernel_2](https://github.com/TheCharlatan/bitcoin/tree/rmNetKernel_2) -> [rmNetKernel_3](https://github.com/TheCharlatan/bitcoin/tree/rmNetKernel_3),  [compare](https://github.com/TheCharlatan/bitcoin/compare/rmNetKernel_2..rmNetKernel_3))

* Addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/28423#discussion_r1321706915), adding a commit for serializing the array as well as a commit for migrating from the raw array to the `std::array` type.
* Addressed @theuni's [comment](https://github.com/bitcoin/bitcoin/pull/28423#pullrequestreview-1620037694), splitting up the commit into the `m_params` member and moving the `GetDefaultPort` functions.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28423#discussion_r1321321923), using the `MessageStartMagic` type directly.

stickies-v commented on src/protocol.h:
```diff
@@ -26,14 +27,14 @@
 {
 public:
-    static constexpr size_t MESSAGE_START_SIZE = 4;
+    using MessageStartChars = MessageStartMagic;
```
nit: I would be okay with dropping this and renaming the 6 places where we use `MessageStartChars`

stickies-v commented on src/serialize.h:
```diff
@@ -203,6 +203,7 @@ template<typename Stream> inline void Serialize(Stream& s, int64_t a ) { ser_wri
 template<typename Stream> inline void Serialize(Stream& s, uint64_t a) { ser_writedata64(s, a); }
 template<typename Stream, int N> inline void Serialize(Stream& s, const char (&a)[N]) { s.write(MakeByteSpan(a)); }
 template<typename Stream, int N> inline void Serialize(Stream& s, const unsigned char (&a)[N]) { s.write(MakeByteSpan(a)); }
+template <typename Stream, typename B, std::size_t N> void Serialize(Stream& s, const std::array<B, N>& a) { (void)/* force byte-type */UCharCast(a.data()); s.write(MakeByteSpan(a)); }
```
nit: ser, that's a long line

TheCharlatan replied:
Seemed awkward to format this line, but leave the others. I think I'll leave this as is.

ajtowns replied:
Might be clearer as `{ s.write(AsBytes(UCharSpanCast(Span(a)))); }` ? I think that has all the same restrictions as you get with the dummy `UCharCast`.

ajtowns replied:
Yeah. Might be another place where C++20 concepts would be good -- `concept BytesLike = requires (B& b) { UCharCast(b) }` and `template<typename Stream, BytesLike B, ..>` or so.

stickies-v commented on src/wallet/db.cpp:
```diff
@@ -136,7 +136,7 @@ bool IsSQLiteFile(const fs::path& path)
 
     // Check the application id matches our network magic
-    return memcmp(Params().MessageStart(), app_id, 4) == 0;
+    return memcmp(Span{Params().MessageStart()}.data(), app_id, 4) == 0;
```
nit: missing include, but do we need span? (+ in `sqlite.cpp`)

```suggestion
    return memcmp(Params().MessageStart().data(), app_id, 4) == 0;
```

stickies-v commented:
re-ACK 4ce0ead928965971d082bfbe2a991703f2f5c827

stickies-v reviewed (approved):
re-ACK 4ce0ead928965971d082bfbe2a991703f2f5c827

TheCharlatan commented:
Updated 4ce0ead928965971d082bfbe2a991703f2f5c827 -> d4170ca919c7f32735774d6f71ad74e04e5fcb25 ([rmNetKernel_3](https://github.com/TheCharlatan/bitcoin/tree/rmNetKernel_3) -> [rmNetKernel_4](https://github.com/TheCharlatan/bitcoin/tree/rmNetKernel_4),  [compare](https://github.com/TheCharlatan/bitcoin/compare/rmNetKernel_3..rmNetKernel_4))

* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28423#discussion_r1323265570), but took another approach by reverting back to the original name. This should keep things consistent and churn low.
* Addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/28423#discussion_r1323294002), removing useless `Span` wrapping. I had the impression this was some minimal required sugar to get msvc to accept the type.

TheCharlatan commented:
Rebased d4170ca919c7f32735774d6f71ad74e04e5fcb25 -> d5067651991f3e6daf456ba13c7036ddc4545352 ([rmNetKernel_4](https://github.com/TheCharlatan/bitcoin/tree/rmNetKernel_4) -> [rmNetKernel_5](https://github.com/TheCharlatan/bitcoin/tree/rmNetKernel_5),  [compare](https://github.com/TheCharlatan/bitcoin/compare/rmNetKernel_4..rmNetKernel_5))

* Fixed silent merge conflict with #28196

stickies-v commented:
re-ACK d506765

hebasto commented on src/serialize.h:
```diff
@@ -301,6 +302,7 @@ template<typename Stream> inline void Unserialize(Stream& s, int64_t& a ) { a =
 template<typename Stream> inline void Unserialize(Stream& s, uint64_t& a) { a = ser_readdata64(s); }
 template<typename Stream, int N> inline void Unserialize(Stream& s, char (&a)[N]) { s.read(MakeWritableByteSpan(a)); }
 template<typename Stream, int N> inline void Unserialize(Stream& s, unsigned char (&a)[N]) { s.read(MakeWritableByteSpan(a)); }
+template <typename Stream, typename B, std::size_t N> void Unserialize(Stream& s, std::array<B, N>& a) { (void)/* force byte-type */UCharCast(a.data()); s.read(MakeWritableByteSpan(a)); }
```
37e2b011136ca1cf00dfb9e575d12f0d035a6a2c

Add `std::array` objects to `serialize_tests.cpp`?

TheCharlatan replied:
Would this be enough for you? Can otherwise do it in a follow-up.
```diff
diff --git a/src/test/serialize_tests.cpp b/src/test/serialize_tests.cpp
index 2f2bb6698c..28d276076e 100644
--- a/src/test/serialize_tests.cpp
+++ b/src/test/serialize_tests.cpp
@@ -87,2 +87,4 @@ BOOST_AUTO_TEST_CASE(sizes)
     BOOST_CHECK_EQUAL(GetSerializeSize(bool(0), 0), 1U);
+    BOOST_CHECK_EQUAL(GetSerializeSize(std::array<uint8_t, 1>{0}, 0), 1U);
+    BOOST_CHECK_EQUAL(GetSerializeSize(std::array<uint8_t, 2>{0, 0}, 0), 2U);
 }
 ```

hebasto replied:
Sorry, I had to be clear in my initial comment that it was a suggestion for a follow-up.

hebasto commented on src/kernel/messagestartchars.h:
```diff
@@ -0,0 +1,13 @@
+#include <array>
+#include <cstdint>
+
+using MessageStartChars = std::array<uint8_t, 4>;
```
Why not `std::byte`?

TheCharlatan replied:
I didn't want to change this data type, since I am not familiar enough with all its users to say what side effects this may have. Also I think `unsigned char` would convey the meaning of the type better than `std::byte`, given that we do use it for string representations.

theuni replied:
Agreed, this PR has crept enough already :)

(Though std::byte makes sense as a follow-up, considering this will likely end up exposed as part of some api)

ajtowns replied:
We already do maths on this in a sense, in that we convert it to an `int32_t` for the `app_id` in wallet/sqlite.cpp. Not seeing what the benefit in switching to `std::byte` here would be.

hebasto reviewed (approved):
ACK d5067651991f3e6daf456ba13c7036ddc4545352.

ajtowns commented on src/protocol.h:
```diff
@@ -26,14 +28,12 @@
 class CMessageHeader
 {
 public:
-    static constexpr size_t MESSAGE_START_SIZE = 4;
```
Defining this as `std::tuple_size_v<MessageStartChars>` seems like it would've touched fewer lines

ajtowns reviewed (commented):
utACK d5067651991f3e6daf456ba13c7036ddc4545352

maflcko reviewed (approved):
left some style nits, better to be left as-is for a follow-up

lgtm ACK d5067651991f3e6daf456ba13c7036ddc4545352 🍛

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: lgtm ACK d5067651991f3e6daf456ba13c7036ddc4545352 🍛
lStXnLnO5f5mgzKS9vTXRcRE1E6bMezm8Whu8+/ejq9XWm3C/d0zha2XRO2Zh+Qup7wt0W56hpPWA2AteHA/CQ==
```

</details>

fanquake commented:
Going to merge this shortly. The 324 integration PR will have to be rebased, but it looks like that will be pushed to again in any case (to address comments). The comments left here can be addressed in a followup PR.

TheCharlatan commented:
Added the following comments to the kernel project tracking issue TODOs:
https://github.com/bitcoin/bitcoin/pull/28423#discussion_r1325169578
https://github.com/bitcoin/bitcoin/pull/28423#discussion_r1324597597
https://github.com/bitcoin/bitcoin/pull/28423#discussion_r1325633787
https://github.com/bitcoin/bitcoin/pull/28423#discussion_r1325636261
https://github.com/bitcoin/bitcoin/pull/28423#discussion_r1325657039
